vmware-event-router/internal/metrics/server.go

```diff
@@ -41,51 +43,87 @@ type Server struct {
 }
 
 // NewServer returns an initialized metrics server binding to addr
-func NewServer(cfg connection.Config) (*Server, error) {
-	var username, password string
-	var basicAuth bool
-
-	addr := cfg.Address
-	switch cfg.Auth.Method {
-	case "basic_auth":
-		basicAuth = true
-		username = cfg.Auth.Secret["username"]
-		password = cfg.Auth.Secret["password"]
-	case "none":
-		break
-	default:
-		return nil, errors.Errorf("unsupported authentication method for metrics server: %q", cfg.Auth.Method)
+func NewServer(cfg *config.MetricsProviderConfigDefault) (*Server, error) {
+	if cfg == nil {
+		return nil, errors.New("no metrics server configuration found")
 	}
 
 	logger := log.New(os.Stdout, color.Teal("[Metrics Server] "), log.LstdFlags)
+	basicAuth := true
+
+	if cfg.Auth == nil || cfg.Auth.BasicAuth == nil {
+		logger.Print("no credentials found, disabling authentication for metrics server")
+		basicAuth = false
+	}
+
 	mux := http.NewServeMux()
+
 	switch basicAuth {
 	case true:
-		mux.Handle(endpoint, withBasicAuth(logger, expvar.Handler(), username, password))
+		mux.Handle(endpoint, withBasicAuth(logger, expvar.Handler(), cfg.Auth.BasicAuth.Username, cfg.Auth.BasicAuth.Password))
 	default:
 		mux.Handle(endpoint, expvar.Handler())
 	}
 
+	err := validateAddress(cfg.BindAddress)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not validate bind address")
+	}
+
 	srv := &Server{
 		http: &http.Server{
-			Addr:         addr,
+			Addr:         cfg.BindAddress,
 			Handler:      mux,
 			ReadTimeout:  httpTimeout,
 			WriteTimeout: httpTimeout,
 		},
 		Logger: logger,
 	}
+
 	return srv, nil
 }
 
+func validateAddress(address string) error {
+	// TODO: this list is not extensive
+	const invalidChars = `abcdefghijklmnopqrstuvwxyz/\ `
+
+	address = strings.ToLower(address)
+	if strings.ContainsAny(address, invalidChars) {
+		return errors.New("invalid character detected")
```

Comment:
I suppose, given the implementation, we don't support DNS in case someone decides to use external metrics server with configured DNS or internal DNS discovery of kubernetes. We can be more explicit with the error here.

Reply:
Right, the `default` is an "embedded" implementation for now. Thus any form of DNS would not work as bind address (which makes sense in most of the cases to avoid the resolver). I will change the error message to make it more meaningful.